internal/profiling/buffer/buffer.go

```diff
@@ -0,0 +1,268 @@
+// optimal number through experiments.
+func floorCPUCount() uint32 {
+	floorExponent := bits.Len32(uint32(runtime.NumCPU())) - 1
+	if floorExponent < 0 {
+		floorExponent = 0
+	}
+	return 1 << uint32(floorExponent)
+}
+
+var numCircularBufferPairs = floorCPUCount()
+
+// CircularBuffer stores the Pushed elements in-memory. It uses a collection of
+// queuePairs to reduce contention on the acquired and written counters within
+// each queuePair; that is, at any given time, there may be floorCPUCount()
+// Pushes happening simultaneously.
+type CircularBuffer struct {
+	mu sync.Mutex
+	qp []*queuePair
+	// qpn is an monotonically incrementing counter that's used to determine
+	// which queuePair a Push operation should write to. This approach's
+	// performance was found to be better than writing to a random queue.
+	qpn    uint32
+	qpMask uint32
+}
+
+var errInvalidCircularBufferSize = errors.New("buffer size is not an exponent of two")
+
+// NewCircularBuffer allocates a circular buffer of size size and returns a
+// reference to the struct. Only circular buffers of size 2^k are allowed
+// (saves us from having to do expensive modulo operations).
+func NewCircularBuffer(size uint32) (*CircularBuffer, error) {
+	if size&(size-1) != 0 {
+		return nil, errInvalidCircularBufferSize
+	}
+
+	n := numCircularBufferPairs
+	if size/numCircularBufferPairs < 8 {
+		// If each circular buffer is going to hold less than a very small number
+		// of items (let's say 8), using multiple circular buffers is very likely
+		// wasteful. Instead, fallback to one circular buffer holding everything.
+		n = 1
+	}
+
+	cb := &CircularBuffer{
+		qp:     make([]*queuePair, n),
+		qpMask: n - 1,
+	}
+
+	for i := uint32(0); i < n; i++ {
+		cb.qp[i] = newQueuePair(size / n)
+	}
+
+	return cb, nil
+}
+
+// Push pushes an element in to the circular buffer.
+func (cb *CircularBuffer) Push(x interface{}) {
+	n := atomic.AddUint32(&cb.qpn, 1) & cb.qpMask
+	q := (*queue)(atomic.LoadPointer(&cb.qp[n].q))
+
+	acquired := atomic.AddUint32(&q.acquired, 1) - 1
+
+	// If drainExecuting is zero, it means that we have incremented acquired
+	// (but not written), so we can guarantee that the Drain operation has not
+	// completed its drainWait. Therefore, it is safe to proceed with the Push
+	// operation on this queue. Otherwise, if drainExecuting is non-zero, it
+	// means that a Drain operation has begun execution, but we don't know how
+	// far along the process it is. If it is past the drainWait check, it is not
+	// safe to proceed with the Push operation. We choose to drop this sample
+	// entirely instead of retrying, as retrying may potentially send the Push
+	// operation into a spin loop (we want to guarantee completion of the Push
+	// operation within a finite time). Before exiting, we increment written so
+	// that any existing drainWaits can proceed.
+	if atomic.LoadUint32(&cb.qp[n].drainExecuting) != 0 {
+		atomic.AddUint32(&q.written, 1)
+		return
```

Comment:
I did change it to a `for { if break }`, but changed it again to this.

No, the caller wouldn't call Push again. This because dropping some pushes is okay if it guarantees finite completion of the push operation. This would fail only if a Drain starts right at the start of a Push, which is pretty rare, but if it happens, we'd still like for profiling to not affect gRPC performance.